// ligdicash/index.js
// Copyright 9-12-2020 (c) Louis Bertson <louisbertson@gmail.com>
//
// Permission is hereby granted, free of charge, to any person obtaining a
// copy of this software and associated documentation files (the
// "Software"), to deal in the Software without restriction, including
// without limitation the rights to use, copy, modify, merge, publish,
// distribute, sublicense, and/or sell copies of the Software, and to permit
// persons to whom the Software is furnished to do so, subject to the
// following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN
// NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM,
// DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE
// USE OR OTHER DEALINGS IN THE SOFTWARE.
const axios = require('axios');
const HEADERS = {};
const commande = {};
commande.invoice = {};
commande.invoice.items = [];
commande.invoice.devise = 'xof';
commande.invoice.description = '';
commande.invoice.customer = '';
commande.invoice.total_amount = '';
commande.invoice.otp = ''; 
commande.store = {};
commande.store.name = 'S.pass Télécom';
commande.website_url = 'spasstelecom.com';
commande.actions = {};
commande.actions.returnUrl = '';
commande.actions.cancelUrl = '';
commande.custom_data = {};
commande.custom_data.boss = 'Nanema';
commande.custom_data.lieu = 'Yako';
HEADERS['User-Agent'] = 'LigdiCash payment  Checkout for Node.js';
HEADERS['Accept'] = 'application/json';
HEADERS['Content-Type'] = 'application/json',
HEADERS['Apikey'] = 'Bearer ZVBGJYG9VS9B283ES';
HEADERS['Authorization'] = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZF9hcHAiOiI0MjQiLCJpZF9hYm9ubmUiOiI2MzQyNiIsImRhdGVjcmVhdGlvbl9hcHAiOiIyMDIwLTEyLTEyIDEyOjE5OjUxIn0.7BkQviVdzbhgSEvA2J-Kog1fKGKaO1_lcfhZ1yAA8Eg';
var self;
function LigdiCash(customer, returnUrl,  cancelUrl, debug) {
	 self =this;
	commande.invoice.customer = customer;
	this.debug = debug || false;
	commande.actions.returnUrl = returnUrl;
	commande.actions.cancelUrl = cancelUrl;
	this.url = 'https://' + (debug ? 'apptest.ligdicash.com' : 'app.ligdicash.com') + '/pay/v01/redirect/checkout-invoice/create';
}
LigdiCash.prototype.post = function(order, controller,callback) {
	   commande.invoice.customer = controller.query.numero;
	   commande.invoice.description = order.note;
	   commande.invoice.total_amount = order.count;
	   for(var i = 0; i<order.items.length;i++){
		   var item = {}
		   item.name = order.items[i].name;
		   item.description = order.items[i].reference;
		   item.quantity = order.items[i].count;
		   item.unit_price = order.items[i].price;
		   item.total_price = order.items[i].count * order.items[i].price;
		   commande.invoice.items.push(item);
	   }
	   var donnee = {};
	   donnee.commande = commande;
	   var data = JSON.stringify(donnee);
    
	axios.post(this.url,data,{headers : HEADERS})
	   .then(function(res){
		   callback(res,null);
		
	   });
	   
	return self;
};


exports.timeout = 10000;
exports.LigdiCash = LigdiCash;
exports.init = function(customer, returnUrl,  cancelUrl, debug) {
	return new LigdiCash(customer, returnUrl,  cancelUrl, debug);
};
exports.create = function(customer, returnUrl,  cancelUrl, debug) {
	return exports.init(customer, returnUrl,  cancelUrl, debug);
};
